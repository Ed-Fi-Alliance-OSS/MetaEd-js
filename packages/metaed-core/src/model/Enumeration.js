// @flow
import type { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import type { EnumerationItem } from './EnumerationItem';
import type { SourceMap } from './SourceMap';
import type { ModelBase } from './ModelBase';

export type EnumerationSourceMap = {
  ...$Exact<TopLevelEntitySourceMap>,
  enumerationItems: Array<SourceMap>,
};

/**
 *
 */
export function newEnumerationSourceMap(): EnumerationSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    enumerationItems: [],
  };
}

export type Enumeration = {
  sourceMap: EnumerationSourceMap,
  ...$Exact<TopLevelEntity>,
  enumerationItems: Array<EnumerationItem>,
};

/**
 *
 */
export function newEnumeration(): Enumeration {
  return {
    ...newTopLevelEntity(),
    type: 'enumeration',
    typeHumanizedName: 'Enumeration',
    enumerationItems: [],
    sourceMap: newEnumerationSourceMap(),
  };
}

/**
 *
 */
export const asEnumeration = (x: ModelBase): Enumeration => ((x: any): Enumeration);
