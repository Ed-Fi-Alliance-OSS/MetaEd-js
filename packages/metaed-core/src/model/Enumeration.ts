import { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import { EnumerationItem } from './EnumerationItem';
import { SourceMap } from './SourceMap';
import { ModelBase } from './ModelBase';

export interface EnumerationSourceMap extends TopLevelEntitySourceMap {
  enumerationItems: SourceMap[];
}

/**
 *
 */
export function newEnumerationSourceMap(): EnumerationSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    enumerationItems: [],
  };
}

export interface Enumeration extends TopLevelEntity {
  sourceMap: EnumerationSourceMap;
  enumerationItems: EnumerationItem[];
}

/**
 *
 */
export function newEnumeration(): Enumeration {
  return {
    ...newTopLevelEntity(),
    type: 'enumeration',
    typeHumanizedName: 'Enumeration',
    enumerationItems: [],
    sourceMap: newEnumerationSourceMap(),
  };
}

/**
 *
 */
export const asEnumeration = (x: ModelBase): Enumeration => x as Enumeration;
