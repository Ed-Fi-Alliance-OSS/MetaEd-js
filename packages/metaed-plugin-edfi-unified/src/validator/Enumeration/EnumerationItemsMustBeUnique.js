// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';
import { failEnumerationItemRedeclarations } from '../ValidatorShared/FailEnumerationItemRedeclarations';

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.enumeration.forEach(enumeration => {
    if (enumeration.enumerationItems.length > 1) {
      failEnumerationItemRedeclarations('EnumerationItemsMustBeUnique', enumeration, enumeration.enumerationItems, failures);
    }
  });

  return failures;
}
