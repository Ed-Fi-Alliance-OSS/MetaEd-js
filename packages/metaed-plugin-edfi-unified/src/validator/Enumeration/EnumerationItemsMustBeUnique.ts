import { MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { failEnumerationItemRedeclarations } from '../ValidatorShared/FailEnumerationItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.enumeration.forEach(enumeration => {
      if (enumeration.enumerationItems.length > 1) {
        failEnumerationItemRedeclarations(
          'EnumerationItemsMustBeUnique',
          enumeration,
          enumeration.enumerationItems,
          failures,
        );
      }
    });
  });

  return failures;
}
