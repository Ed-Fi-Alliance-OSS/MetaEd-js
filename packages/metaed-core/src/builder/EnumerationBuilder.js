// @flow
import type { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newEnumeration, asEnumeration } from '../model/Enumeration';
import { newSchoolYearEnumeration } from '../model/SchoolYearEnumeration';
import type { EnumerationSourceMap } from '../model/Enumeration';
import type { EnumerationItem } from '../model/EnumerationItem';
import { newEnumerationItem, NoEnumerationItem } from '../model/EnumerationItem';
import { extractDocumentation, extractShortDescription, squareBracketRemoval, isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import type { MetaEdEnvironment } from '../MetaEdEnvironment';
import type { ValidationFailure } from '../validator/ValidationFailure';
import { sourceMapFrom } from '../model/SourceMap';

export class EnumerationBuilder extends TopLevelEntityBuilder {
  currentEnumerationItem: EnumerationItem;

  constructor(metaEd: MetaEdEnvironment, validationFailures: Array<ValidationFailure>) {
    super(metaEd, validationFailures);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.enteringEntity(newEnumeration);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
        namespaceInfo: this.currentTopLevelEntity.namespaceInfo.sourceMap.type,
      });
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.exitingEntity();
  }

  enterEnumerationName(context: MetaEdGrammar.EnumerationNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    const enumerationName = context.ID().getText();

    // need to differentiate SchoolYear from other enumerations - overwrite with new type
    if (enumerationName === 'SchoolYear') {
      this.enteringEntity(newSchoolYearEnumeration);
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
        namespaceInfo: this.currentTopLevelEntity.namespaceInfo.sourceMap.type,
      });
    }

    this.enteringName(enumerationName);
    Object.assign(this.currentTopLevelEntity.sourceMap, {
      metaEdName: sourceMapFrom(context),
    });
  }

  enterEnumerationItemDocumentation(context: MetaEdGrammar.EnumerationItemDocumentationContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.documentation = extractDocumentation(context);
    this.currentEnumerationItem.sourceMap.documentation = sourceMapFrom(context);
  }

  enterEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    this.currentEnumerationItem = newEnumerationItem();
    this.currentEnumerationItem.namespaceInfo.sourceMap = this.currentTopLevelEntity.namespaceInfo.sourceMap;
    this.currentEnumerationItem.sourceMap.type = sourceMapFrom(context);
  }

  exitEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity || this.currentEnumerationItem === NoEnumerationItem) return;
    asEnumeration(this.currentTopLevelEntity).enumerationItems.push(this.currentEnumerationItem);
    ((this.currentTopLevelEntity.sourceMap: any): EnumerationSourceMap).enumerationItems.push(sourceMapFrom(context));
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterMetaEdId(context: MetaEdGrammar.MetaEdIdContext) {
    if (context.exception || context.METAED_ID() == null || context.METAED_ID().exception != null || isErrorText(context.METAED_ID().getText())) return;
    if (this.currentEnumerationItem !== NoEnumerationItem) {
      this.currentEnumerationItem.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
      this.currentEnumerationItem.sourceMap.metaEdId = sourceMapFrom(context);
    } else {
      super.enterMetaEdId(context);
    }
  }

  enterShortDescription(context: MetaEdGrammar.ShortDescriptionContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.shortDescription = extractShortDescription(context);
    this.currentEnumerationItem.sourceMap.shortDescription = sourceMapFrom(context);
  }
}
