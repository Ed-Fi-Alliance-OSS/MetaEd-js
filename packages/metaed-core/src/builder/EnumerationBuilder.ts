import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newEnumeration, asEnumeration } from '../model/Enumeration';
import { newSchoolYearEnumeration } from '../model/SchoolYearEnumeration';
import { EnumerationSourceMap } from '../model/Enumeration';
import { EnumerationItem } from '../model/EnumerationItem';
import { newEnumerationItem, NoEnumerationItem } from '../model/EnumerationItem';
import { extractDocumentation, extractShortDescription, isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { MetaEdEnvironment } from '../MetaEdEnvironment';
import { ValidationFailure } from '../validator/ValidationFailure';
import { sourceMapFrom } from '../model/SourceMap';

/**
 * An ANTLR4 listener that creates Enumeration entities.
 */
export class EnumerationBuilder extends TopLevelEntityBuilder {
  currentEnumerationItem: EnumerationItem;

  constructor(metaEd: MetaEdEnvironment, validationFailures: ValidationFailure[]) {
    super(metaEd, validationFailures);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.enteringEntity(newEnumeration);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
      });
    }
  }

  exitEnumeration(_context: MetaEdGrammar.EnumerationContext) {
    this.exitingEntity();
  }

  enterEnumerationName(context: MetaEdGrammar.EnumerationNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    const enumerationName = context.ID().getText();

    // need to differentiate SchoolYear from other enumerations - overwrite with new type
    if (enumerationName === 'SchoolYear') {
      this.enteringEntity(newSchoolYearEnumeration);
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
      });
    }

    this.enteringName(enumerationName);
    Object.assign(this.currentTopLevelEntity.sourceMap, {
      metaEdName: sourceMapFrom(context),
    });
  }

  enterEnumerationItemDocumentation(context: MetaEdGrammar.EnumerationItemDocumentationContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.documentation = extractDocumentation(context);
    this.currentEnumerationItem.sourceMap.documentation = sourceMapFrom(context);
  }

  enterEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    this.currentEnumerationItem = newEnumerationItem();
    this.currentEnumerationItem.sourceMap.type = sourceMapFrom(context);
  }

  exitEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity || this.currentEnumerationItem === NoEnumerationItem) return;
    asEnumeration(this.currentTopLevelEntity).enumerationItems.push(this.currentEnumerationItem);
    (this.currentTopLevelEntity.sourceMap as EnumerationSourceMap).enumerationItems.push(sourceMapFrom(context));
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterShortDescription(context: MetaEdGrammar.ShortDescriptionContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.shortDescription = extractShortDescription(context);
    this.currentEnumerationItem.sourceMap.shortDescription = sourceMapFrom(context);
  }
}
