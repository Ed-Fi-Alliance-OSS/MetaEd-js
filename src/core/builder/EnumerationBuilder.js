// @flow
import { MetaEdGrammar } from '../../grammar/gen/MetaEdGrammar';
import TopLevelEntityBuilder from './TopLevelEntityBuilder';
import { enumerationFactory } from '../model/Enumeration';
import { schoolYearEnumerationFactory } from '../model/SchoolYearEnumeration';
import type { Enumeration } from '../model/Enumeration';
import type { EnumerationItem } from '../model/EnumerationItem';
import { enumerationItemFactory, NoEnumerationItem } from '../model/EnumerationItem';
import { extractDocumentation, extractShortDescription, squareBracketRemoval } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { EntityRepository } from '../model/Repository';

export default class EnumerationBuilder extends TopLevelEntityBuilder {
  currentEnumerationItem: EnumerationItem;

  constructor(repository: EntityRepository) {
    super(repository);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  // eslint-disable-next-line no-unused-vars
  enterEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.enteringEntity(enumerationFactory);
  }

  // eslint-disable-next-line no-unused-vars
  exitEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.exitingEntity();
  }

  enterEnumerationName(context: MetaEdGrammar.EnumerationNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception) return;
    const enumerationName = context.ID().getText();

    // need to differentiate SchoolYear from other enumerations - overwrite with new type
    if (enumerationName === 'SchoolYear') {
      this.enteringEntity(schoolYearEnumerationFactory);
    }

    this.enteringName(enumerationName);
  }

  enterEnumerationItemDocumentation(context: MetaEdGrammar.EnumerationItemDocumentationContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.documentation = extractDocumentation(context);
  }

  // eslint-disable-next-line no-unused-vars
  enterEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    this.currentEnumerationItem = enumerationItemFactory();
  }

  // eslint-disable-next-line no-unused-vars
  exitEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity || this.currentEnumerationItem === NoEnumerationItem) return;
    ((this.currentTopLevelEntity: any): Enumeration).enumerationItems.push(this.currentEnumerationItem);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterMetaEdId(context: MetaEdGrammar.MetaEdIdContext) {
    if (context.METAED_ID() == null || context.METAED_ID().exception != null) return;
    if (this.currentEnumerationItem !== NoEnumerationItem) {
      this.currentEnumerationItem.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
    } else {
      super.enterMetaEdId(context);
    }
  }

  enterShortDescription(context: MetaEdGrammar.ShortDescriptionContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.shortDescription = extractShortDescription(context);
  }
}
