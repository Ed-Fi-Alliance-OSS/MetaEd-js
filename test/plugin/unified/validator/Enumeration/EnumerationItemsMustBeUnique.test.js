// @noflow
import EnumerationBuilder from '../../../../../src/core/builder/EnumerationBuilder';
import MetaEdTextBuilder from '../../../../core/MetaEdTextBuilder';
import { repositoryFactory } from '../../../../../src/core/model/Repository';
import type { Repository } from '../../../../../src/core/model/Repository';
import { validate } from '../../../../../src/plugin/unified/validator/Enumeration/EnumerationItemsMustBeUnique';
import type { ValidationFailure } from '../../../../../src/core/validator/ValidationFailure';

describe('when enumeration items have different short descriptions', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartEnumeration('EntityName')
      .withDocumentation('EntityDocumentation')
      .withEnumerationItem('ShortDescription1', 'EnumerationItemDocumentation1')
      .withEnumerationItem('ShortDescription2', 'EnumerationItemDocumentation2')
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(new EnumerationBuilder(repository.entity, [], new Map()));

    failures = validate(repository);
  });

  it('should build one enumeration', () => {
    expect(repository.entity.enumeration.size).toBe(1);
  });


  it('should have no validation failures()', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when enumeration items have duplicate short descriptions', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartEnumeration('EntityName')
      .withDocumentation('EntityDocumentation')
      .withEnumerationItem('ShortDescription', 'EnumerationItemDocumentation1')
      .withEnumerationItem('ShortDescription', 'EnumerationItemDocumentation2')
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(new EnumerationBuilder(repository.entity, [], new Map()));

    failures = validate(repository);
  });

  it('should build one enumeration', () => {
    expect(repository.entity.enumeration.size).toBe(1);
  });

  it('should have validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('EnumerationItemsMustBeUnique');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when enumeration items have duplicate short descriptions should have validation failure -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when enumeration items have duplicate short descriptions should have validation failure -> sourceMap');
  });
});

describe('when enumeration items have multiple duplicate short descriptions', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartEnumeration('EntityName')
      .withDocumentation('EntityDocumentation')
      .withEnumerationItem('ShortDescription1', 'EnumerationItemDocumentation1')
      .withEnumerationItem('ShortDescription1', 'EnumerationItemDocumentation2')
      .withEnumerationItem('ShortDescription2', 'EnumerationItemDocumentation1')
      .withEnumerationItem('ShortDescription2', 'EnumerationItemDocumentation2')
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(new EnumerationBuilder(repository.entity, [], new Map()));

    failures = validate(repository);
  });

  it('should build one enumeration', () => {
    expect(repository.entity.enumeration.size).toBe(1);
  });

  it('should have validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('EnumerationItemsMustBeUnique');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when enumeration items have multiple duplicate short descriptions should have validation failure -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when enumeration items have multiple duplicate short descriptions should have validation failure -> sourceMap');
  });
});
