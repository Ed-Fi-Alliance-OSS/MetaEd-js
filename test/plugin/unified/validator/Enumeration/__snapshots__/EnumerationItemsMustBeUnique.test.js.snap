// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when enumeration items have duplicate short descriptions should have validation failure -> message 1`] = `"Enumeration EntityName declares duplicate item ShortDescription. "`;

exports[`when enumeration items have duplicate short descriptions should have validation failure -> sourceMap 1`] = `
Object {
  "column": 2,
  "line": 2,
  "tokenText": "Enumeration",
}
`;

exports[`when enumeration items have multiple duplicate short descriptions should have validation failure -> message 1`] = `"Enumeration EntityName declares duplicate items ShortDescription1, ShortDescription2. "`;

exports[`when enumeration items have multiple duplicate short descriptions should have validation failure -> sourceMap 1`] = `
Object {
  "column": 2,
  "line": 2,
  "tokenText": "Enumeration",
}
`;
